### PR TITLE
feat(integ): add secrets management identity registration settings tests

### PR DESCRIPTION
**Pull request by jericht**
### Notes
- Adds integration tests for the Secrets Management identity registration settings feature implemented in https://github.com/aws/aws-rfdk/pull/576
    - The `deadline_05_secretsManagement` test suite now uses a default (dummy) UBL certificate secret (create if not found). This secret is meant to be created once per account + region these integ tests are run in and referenced in all future integ tests, since its content is static (basically empty).
- Adds a new developer mode to the integration test script so that the CF stacks will not be torn down, and making the bastion host accessible via SSM, making it easier to iterate on integ test development
- Reorganizes the `NetworkTier` to have dedicated subnet configurations for each Deadline component (RQ ALB, workers, license forwarder, etc.)
- Adds options to the `RenderStruct` construct to setup a `UsageBasedLicensing` construct. This is currently only used in the `deadline_05_secretsManagement` test suite.
- Adds a new `SepWorkerStruct` that configures the Spot Event Plugin via `ConfigureSpotEventPlugin` construct

~~**Note for reviewers:** This is based off of the changes in https://github.com/aws/aws-rfdk/pull/576. It will be easier to review by looking at the commits authored by me so only my changes are shown in the file diffs.~~

### Testing
Took the following steps to test:
- Staged Deadline 10.1.18 Docker recipes & updated them with the changes for the Secrets Management feature
- Changed the `SecretsManagementIdentityRegistration` construct to use a version of Deadline Client with the Secrets Management changes
- Let the integ test create a dummy secret for the UBL certificates, then ran the integration tests and verified they passed
- Ran all the integ test suites and verified they passed

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: integ/components/_infrastructure/lib/network-tier.ts
Comment (horsmand) [inline integ/components/_infrastructure/lib/network-tier.ts]: Do we need this enum? I don't see it referenced anywhere other than in the type right below it and we could just list the subnets there.

File: integ/components/_infrastructure/lib/network-tier.ts
Comment (horsmand) [inline integ/components/_infrastructure/lib/network-tier.ts]: What was the reasoning to use a `type` over an `interface` here? I'm not sure how this plays with JSII. That probably doesn't matter since I'm quite sure we never transpile this integ package, but it's probably best if we follow the typescript restrictions everywhere in our repository for consistency. 

From https://aws.github.io/jsii/user-guides/lib-author/typescript-restrictions/#interfaces:
> The jsii type model distinguishes two kinds of interfaces:
>
> * Behavioral interfaces, which can declare methods and properties
> * Structs, which are immutable pure data entities, and can consequently only declare readonly properties
>
> A name convention is used to distinguish between these two: behavioral interfaces must have a name that starts with a I prefix, while structs must not have such a prefix.

I don't see why we can't use the struct interface here.

File: integ/scripts/bash/rfdk-integ-e2e.sh
Comment (horsmand) [inline integ/scripts/bash/rfdk-integ-e2e.sh:100]: I can't seem to comment on lines that aren't modified in this commit, but above this on line 88 there's a check that is broken. It currently has:
```
# If executing worker fleet tests, find Deadline AMIs based on supplied version
if [ ! "${SKIP_deadline_03_repository_TEST-}" = true ]; then
    source $BASH_SCRIPTS/fetch-worker-amis.sh
fi
```
I think we want to make sure we're not skipping `deadline_03_workerFleetHttp`, `deadline_03_workerFleetHttps`, and `deadline_05_secretsManagement`. I almost wonder if the check is worth doing since it's easy to get missed when we make modifications and fetching AMIs likely isn't too expensive.

**Comment by horsmand** [comment]
This commit just got merged to the feature branch so we'll need to rebase on it and set the credentials removal policy to `DESTROY` https://github.com/aws/aws-rfdk/commit/d0241d6989b1af1fbced953ed91444946b2655f8#diff-299db76f4b1d275d1140f5b68b75b5954a2c54ef4f6770dfc410bb21668aea8eR318

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (horsmand) [inline integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts]: Should this return undefined rather than throw an error? I don't see how these tests work without having the UBL env vars defined, but the documentation says it should be possible.

File: integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts
Comment (horsmand) [inline integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts]: These don't seem optional

File: integ/test-config.sh
Comment (horsmand) [inline integ/test-config.sh:34]: How is this intended to be used? Will UBL be used in our canaries or pipelines?

File: integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts
Comment (horsmand) [inline integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts]: What is the test instance connecting to these fleets for?

File: integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts
Comment (horsmand) [inline integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts]:
```suggestion
   * Generates CfnOutputs for the Render Queue ALB subnets and Deadline client subnets.
```

File: integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts
Comment (horsmand) [inline integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts]: Isn't UBL optional?

**Review by horsmand — CHANGES_REQUESTED** [review]
Overall this looks good, I really like that you fixed up the `--deploy-and-test-only` flag by adding the `DEV_MODE` env var. The only part I'm unsure about is the intention of adding UBL and whether it's actually optional given how it's currently written.

File: integ/components/_infrastructure/lib/network-tier.ts
Comment (jericht) [reply integ/components/_infrastructure/lib/network-tier.ts]: We don't. I originally thought it would be needed but turned out it wasn't used, so I've removed it

File: integ/components/_infrastructure/lib/network-tier.ts
Comment (jericht) [reply integ/components/_infrastructure/lib/network-tier.ts]: Changed to a struct interface

File: integ/test-config.sh
Comment (jericht) [reply integ/test-config.sh:34]: Sorry, forgot to put a comment here. Originally, I intended for this to be required when running the Secrets Management tests, but it makes more sense to make them optional overrides and have default behavior (e.g. upload dummy secret and use dummy license limits). I've added a comment here as well as implemented the default behavior.

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts]: I changed this to create a default dummy secret for UBL when this in not defined

File: integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts]: These are actually not needed, removed

File: integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts]: It's not, I forgot to remove this. Removed

File: integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/lib/secretsManagement-testing-tier.ts]: Updated

File: integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts]: I've changed the UBL stuff to always get setup. This way our SM integ tests always have the same coverage, regardless of whether UBL certs were provided or not.

File: integ/scripts/bash/rfdk-integ-e2e.sh
Comment (jericht) [reply integ/scripts/bash/rfdk-integ-e2e.sh:100]: Yeah I agree, I've removed that if condition.

File: integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts
Comment (horsmand) [inline integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts]: I'm a little confused by the use of arrays and this `testSuiteId` in this file. I don't see the value of `testSuiteId` changing anywhere.

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (horsmand) [inline integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts]: It looks like we've added a new dependency just to create this empty zip file. Why not just keep an empty zip file in our repo that we can use? Binary files in a git repo are generally bad, but I feel like this might warrant an exception since the file size is negligible. What do you think?

**Comment by horsmand** [comment]
I forgot to mention that my last review was based on this diff, which covers the 2 latest commits: https://github.com/aws/aws-rfdk/pull/596/files/4bb8df5b36d32f0c435466b64cf21231950086d9..0a58b66e61cdca280e35e2d907b099ce7f485956

File: integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/test/deadline_05_secretsManagement.test.ts]: Good point. Unlike the other component tests, this test only has one test suite, I've removed this.

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts]: Good point! I think having the static empty file is much cleaner, updated.

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (kozlove-aws) [inline integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts]: if we expect here zero or negative values could be worth to validate if value more then `UsageBasedLicense.UNLIMITED`

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (kozlove-aws) [inline integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts:142]: How can we be sure that the only possible error is when the secret is not found? 
Could we check error type here?

**Review by kozlove-aws — COMMENTED** [review]
Looks good, just a couple tiny things.

File: integ/lib/storage-struct.ts
Comment (horsmand) [inline integ/lib/storage-struct.ts:83]: Let's add a comment about why we need to create this all the time now but not the DocDB

**Review by horsmand — CHANGES_REQUESTED** [review]
We're very close! Just one comment for clarification.

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (horsmand) [reply integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts]: I know I saw it explained somewhere else in this PR, but maybe we can just add a comment here to say that anything that isn't a positive integer will be treated as unlimited.

Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts]: Good call out! I think we can leave this as is though. Technically, there is no upper limit on the license count limit (see our [license-forwarder recipe documentation for UBL_LIMITS](https://gallery.ecr.aws/aws-thinkbox/deadline/license-forwarder)), and if there is a max value in Deadline, I think we can let Deadline enforce that. How does that sound?

File: integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts
Comment (jericht) [reply integ/components/deadline/deadline_05_secretsManagement/bin/deadline_05_secretsManagement.ts:142]: Good idea, added a check to make sure we get the `ResourceNotFoundException` error code

File: integ/lib/storage-struct.ts
Comment (jericht) [reply integ/lib/storage-struct.ts:83]: Sure, added a little comment here as well as over the relevant `vpcSubnets` field

Comment (jericht) [reply integ/lib/storage-struct.ts:83]: Also just to make sure we're on the same page here, we can't make the DocDB all the time since the `StorageStruct` switches the type of database it uses based on the props it's given (see the `if` statements below this code), but it does not switch the filesystem type, which is why I have refactored it out.

Previous behavior of filesystem creation w.r.t the `databaseType` prop was:
- `DatabaseType.DocDB` => Create EFS and pass it into the `Repository`
- `DatabaseType.MongoDB` => Let `Repository` create the EFS
- `undefined` => Also let `Repository` create the EFS

I found it strange that we only created the EFS in the DocDB case since those two things aren't really tied to each other for our integ tests. Since I needed something other than the default EFS created by our `Repository`, I decided to just refactor the EFS creation outside of the database creation branches.

**Comment by jericht** [comment]
@kozlove-aws @horsmand This is ready for re-review. Please don't merge if both of you approve, I'm going to do a final run of the integ test and merge it myself once it passes.

**Review by horsmand — COMMENTED** [review]
Ready to approve after we confirm the latest run of the tests pass.

**Review by kozlove-aws — APPROVED** [review]

**Comment by jericht** [comment]
@horsmand The tests just passed, this is good to go

**Review by horsmand — APPROVED** [review]

**Pull request merged**
